src/bpftrace.cpp

```diff
@@ -336,19 +334,13 @@ std::vector<uint64_t> BPFtrace::get_arg_values(std::vector<Field> args, uint8_t*
         arg_values.push_back((uint64_t)(arg_data+arg.offset));
         break;
       case Type::sym:
-        resolved_symbols.emplace_back(strdup(
-              resolve_sym(*(uint64_t*)(arg_data+arg.offset)).c_str()));
-        arg_values.push_back((uint64_t)resolved_symbols.back().get());
+        arg_values.push_back((uint64_t)strdup(resolve_sym(*(uint64_t*)(arg_data+arg.offset)).c_str()));
```

Comment:
Trying to reason through this (still learning C++).
I think it works, but only by causing a memory leak.

```c++
resolve_sym(/* .. */)
```
Returns a _temporary_ std::string.

```c++
/* std::string */.c_str()
```
Gets a [raw pointer to the internal buffer](https://stackoverflow.com/a/17403222/5257399) inside the std::string.

We can't interact with the std::string any more, but it is not destroyed yet; it [lives until the end of the full expression](https://stackoverflow.com/a/2506800/5257399).
So our raw-pointer still points to valid memory.
But we need to make it live a bit longer.

```c++
strdup(/* char* */)
```
[Mallocs a duplicate](https://stackoverflow.com/a/252802/5257399) of the C-style string.
This memory won't get destroyed when the expression ends.
**warning: memory leak; we have no plan to dealloc this string**.

We should make a scoped/unique pointer of this raw pointer, so that the duplicated string lives long enough that we can `printf()` the contents of `arg_values`, yet gets cleared on exiting the function.

The unique_ptr would need to be stack-allocated (otherwise it's a temporary value, to which all references are lost when we reach the end of the full-expression).
Sadly we've come full-circle. The `std::vector<std::unique_ptr<char>>` is necessary.

Reply:
See #194